Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Serializer.dfy

```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT 
+ *******************************************************************************/
+
+/**
+ Implements high-level serialization (JSON values to utf-8 bytes).
+ */
+module DafnyStdLibs.JSON.Serializer {
+  import Collections.Seqs
+  import Math
+  import Values
+  import Spec
+  import opened Wrappers
+  import opened BoundedInts
+  import opened Strings
+  import opened Errors
+  import opened DynamicArray
+  import opened Grammar
+  import opened Utils.Views.Core
+
+  type Result<+T> = SerializationResult<T>
+  type bytes = seq<uint8>
+  type bytes32 = bs: bytes | |bs| < TWO_TO_THE_32
+  type string32 = s: string | |s| < TWO_TO_THE_32
+
+  function Bool(b: bool): jbool {
+    View.OfBytes(if b then TRUE else FALSE)
+  }
+
+  function CheckLength<T>(s: seq<T>, err: SerializationError): Outcome<SerializationError> {
+    Outcome.Need(|s| < TWO_TO_THE_32, err)
+  }
+
+  function String(str: string): Result<jstring> {
+    var bs :- Spec.EscapeToUTF8(str);
+    var o := CheckLength(bs, StringTooLong(str));
+    if o.Pass? then
+      Success(Grammar.JString(Grammar.DOUBLEQUOTE, View.OfBytes(bs), Grammar.DOUBLEQUOTE))
+    else
+      Failure(o.error)
+  }
+
+  function Sign(n: int): jminus {
+    View.OfBytes(if n < 0 then ['-' as byte] else [])
+  }
+
+  module ByteStrConversion refines Strings.ParametricConversion {
```

Comment:
Oh but there's also two copies of it? Okay let's move it to Strings before merging unless that's really difficult for some reason. :)

Reply:
Ran into an issue when trying to move it into `Strings`, thus created a JSON submodule `JSON.ByteStrConversion`.